FirebaseAuth/Sources/Auth/FIRAuth.m

```diff
@@ -500,70 +506,88 @@ - (nullable instancetype)initWithAPIKey:(NSString *)APIKey
     [GULSceneDelegateSwizzler proxyOriginalSceneDelegate];
 #endif  // TARGET_OS_IOS
 
-    // Continue with the rest of initialization in the work thread.
-    __weak FIRAuth *weakSelf = self;
-    dispatch_async(FIRAuthGlobalWorkQueue(), ^{
-      // Load current user from Keychain.
-      FIRAuth *strongSelf = weakSelf;
-      if (!strongSelf) {
-        return;
-      }
-      NSString *keychainServiceName =
-          [FIRAuth keychainServiceNameForAppName:strongSelf->_firebaseAppName];
-      if (keychainServiceName) {
-        strongSelf->_keychainServices =
-            [[FIRAuthKeychainServices alloc] initWithService:keychainServiceName];
-        strongSelf.storedUserManager =
-            [[FIRAuthStoredUserManager alloc] initWithServiceName:keychainServiceName];
-      }
+    if ([application isProtectedDataAvailable]) {
+      [self protectedDataInitialization:application];
+    } else {
+      // Add listener for UIApplicationProtectedDataDidBecomeAvailable.
+      self->_protectedDataDidBecomeAvailableObserver = [[NSNotificationCenter defaultCenter]
+          addObserverForName:UIApplicationProtectedDataDidBecomeAvailable
+                      object:nil
+                       queue:nil
+                  usingBlock:^(NSNotification *notification) {
+                    [self protectedDataInitialization:application];
+                  }];
+    }
+  }
+  return self;
+}
 
-      NSError *error;
-      NSString *storedUserAccessGroup =
-          [strongSelf.storedUserManager getStoredUserAccessGroupWithError:&error];
-      if (!error) {
-        if (!storedUserAccessGroup) {
-          FIRUser *user;
-          if ([strongSelf getUser:&user error:&error]) {
-            strongSelf.tenantID = user.tenantID;
-            [strongSelf updateCurrentUser:user byForce:NO savingToDisk:NO error:&error];
-            self->_lastNotifiedUserToken = user.rawAccessToken;
-          } else {
-            FIRLogError(kFIRLoggerAuth, @"I-AUT000001",
-                        @"Error loading saved user when starting up: %@", error);
-          }
+- (void)protectedDataInitialization:(UIApplication *)application {
```

Comment:
Since UIApplication isn't available on macOS, you may need to `#if TARGET_OS` this to provide two separate functions, one that takes a parameter and one that doesn't.

Reply:
Done -- did some refactoring to get around this.